### ISSUE TITLE
ModelCheckpoint Callback for Best Model Preservation

### ISSUE THREAD

**Issue by celsofranssa** [sdv-dev/SDV#2397]
**Description:**

Introduce a `ModelCheckpoint` callback within the synthesizer's training framework to automatically save the model's weights or the entire model at specified intervals. This feature ensures that the best-performing model, based on a user-defined metric, is preserved during training.

**Justification:**

Incorporating a `ModelCheckpoint` callback aligns with standard practices in machine learning frameworks, such as PyTorch Lightning, where model checkpoints are utilized to save models or weights at specified intervals, facilitating the preservation of the best-performing models during training. This functionality is crucial for scenarios where training is computationally intensive, and retaining the best model is essential for future inference or fine-tuning.

**Proposed Implementation:**

- **Callback Integration:** Implement a callback system in the synthesizer's `fit` method, allowing users to specify a `ModelCheckpoint` callback.
- **Monitoring Metrics:** Enable the callback to monitor user-specified metrics (e.g., validation loss, accuracy) to determine the best model.
- **Saving Criteria:** Provide options to save the model at the end of every epoch or only when the monitored metric shows improvement.
- **File Management:** Allow customization of the checkpoint file path and naming conventions, including options to save only the model weights or the entire model architecture.

**Example Usage:**

```python
# Define the ModelCheckpoint callback
checkpoint_callback = ModelCheckpoint(
    dirpath='checkpoints/',
    filename='best_model-{epoch:02d}-{val_loss:.2f}',
    monitor='val_loss',
    mode='min',
    save_top_k=1,
    save_weights_only=True
)

# Train the synthesizer with the callback
synthesizer.fit(train_data, callbacks=[checkpoint_callback])
```

In this example, the `ModelCheckpoint` callback saves the model to the specified directory with a filename that includes the epoch number and validation loss, whenever there is an improvement in validation loss during training.

Implementing this feature would enhance the robustness and usability of the synthesizer by ensuring that the best model obtained during training is readily available for deployment or further training.
